src/stores/RoomListStore.js

```diff
@@ -426,10 +426,28 @@ class RoomListStore extends Store {
         let foundBoundary = false;
         let pushedEntry = false;
 
+        // we won't have reached this point with a manually sorted tag so expect only Alphabetic and Recents sorting
+        const algorithm = getListAlgorithm(tag, this._state.algorithm);
+
+        let comparator;
+        switch (algorithm) {
+            case ALGO_RECENT:
+                comparator = (entryA, entryB) => this._recentsComparator(entryA, entryB, lastTimestampFn);
+                break;
+            case ALGO_ALPHABETIC:
+                comparator = this._lexicographicalComparator;
+                break;
+            default:
+                comparator = (a, b) => lastTimestampFn(a.room) >= lastTimestampFn(b.room);
+        }
+
+        let existingIndex = null; // remove this at the end to not mess up ordering
+        let i = -1;
         for (const entry of existingEntries) {
+            i++;
             // We insert our own record as needed, so don't let the old one through.
-            if (entry.room.roomId === room.roomId) {
-                continue;
+            if (existingIndex === null && entry.room.roomId === room.roomId) {
```

Comment:
This is probably where your duplication is happening: multiple copies of a room can be sent through this function, which is why it does an early filter rather than selective removal.